editor/src/core/shared/github/helpers.ts

```diff
@@ -855,53 +856,111 @@ export function mergeProjectContentsTree(
-          githubUserDetails,
-          lastRefreshedCommit,
-          originCommitSha,
-          operationContext,
-          'polling',
-        )
-      } finally {
-        // Trigger another one to run Xms _after_ this has finished.
-        githubPollingTimeoutId = window.setTimeout(pollGithub, GITHUB_REFRESH_INTERVAL_MILLISECONDS)
-      }
+type GithubAuthState = 'not-authenticated' | 'missing-user-details' | 'ready'
+
+export function useGithubPolling() {
+  const dispatch = useDispatch()
+
+  const [tick, setTick] = React.useState(0)
+  const [lastTick, setLastTick] = React.useState<number | null>(null)
+  const [timeoutId, setTimeoutId] = React.useState<number | null>(null)
+
+  const githubAuthenticated = useEditorState(
+    Substores.userState,
+    (store) => store.userState.githubState.authenticated,
+    'useGithubPolling githubAuthenticated',
+  )
+  const githubData = useEditorState(
+    Substores.github,
+    (store) => ({
+      githubUserDetails: store.editor.githubData.githubUserDetails,
+      lastRefreshedCommit: store.editor.githubData.lastRefreshedCommit,
+      targetRepository: store.editor.githubSettings.targetRepository,
+      branchName: store.editor.githubSettings.branchName,
+      originCommit: store.editor.githubSettings.originCommit,
+    }),
+    'useGithubPolling githubData',
+  )
+  const branchOriginContentsChecksums = useEditorState(
+    Substores.derived,
+    (store) => store.derived.branchOriginContentsChecksums,
+    'useGithubPolling branchOriginContentsChecksums',
+  )
+
+  const refreshAndScheduleGithubData = React.useCallback(async () => {
+    await refreshGithubData(
+      dispatch,
+      githubData.targetRepository,
+      githubData.branchName,
+      branchOriginContentsChecksums,
+      githubData.lastRefreshedCommit,
+      githubData.originCommit,
+      githubOperationContext(),
+      'polling',
+    )
+
+    // schedule a new tick for the next Xms
+    setTimeoutId(
+      window.setTimeout(() => {
+        setTick((t) => t + 1)
+      }, GITHUB_REFRESH_INTERVAL_MILLISECONDS),
+    )
+  }, [dispatch, branchOriginContentsChecksums, githubData])
+
+  const authState = React.useMemo((): GithubAuthState => {
+    return !githubAuthenticated
+      ? 'not-authenticated'
+      : githubData.githubUserDetails == null
+      ? 'missing-user-details'
+      : 'ready'
+  }, [githubAuthenticated, githubData])
+
+  // react to auth state and tick changes
+  React.useEffect(() => {
+    switch (authState) {
+      case 'not-authenticated':
+        dispatch([updateGithubData(emptyGithubData())])
+        break
+      case 'missing-user-details':
+        void GithubHelpers.getUserDetailsFromServer().then((userDetails) => {
+          dispatch(
+            userDetails == null
+              ? resetGithubStateAndDataActions()
+              : updateUserDetailsFromServerActions(userDetails),
+          )
+        })
+        break
+      case 'ready':
+        if (lastTick !== tick) {
```

Comment:
Should it be ordered like so?
```suggestion
        if (lastTick < tick) {
```

Reply:
I _think_ so yeah